File: problems/70. Climbing Stairs/solution.py
class Solution:
    def climbStairs(self, n: int) -> int:
        return self.helper(n, {1: 1, 2: 2})

    def helper(self, n, lookup):
        if n not in lookup:
            lookup[n] = self.helper(n-1, lookup) + self.helper(n-2, lookup)
        return lookup[n]

File: problems/70. Climbing Stairs/solution.cpp
class Solution
{
public:
    int climbStairs(int n)
    {
        unordered_map<int, int> cache{{1, 1}, {2, 2}};
        return recursive(n, cache);
    }

    int recursive(int n, unordered_map<int, int> &cache)
    {

        if (cache.find(n) == cache.end())
        {
            cache[n] = recursive(n - 1, cache) + recursive(n - 2, cache);
        }
        return cache[n];
    }
};